### PR TITLE
[DNM] bluestore: adaptive throttle tied to latency

### PR DESCRIPTION
**Pull request by ivancich**
To turn on feature, you'll need to add to your ceph.conf file:
```
bluestore latency managed throttle = true
```

Other configuration options include bluestore_throttle_latency_min, bluestore_throttle_latency_max, and bluestore_throttle_latency_window_msec.

The general strategy is to look at the commit latency every `bluestore_throttle_latency_window_msec` milliseconds. If within five samples three show samples above `bluestore_throttle_latency_max`, increase the throttle. Likewise if there are three samples below `bluestore_throttle_latency_min`, decrease the throttle. The throttle is increased/decreased to values that are powers of the square root of 2 (e.g., 8192, 11585, 16384, 23170, 32768, 46341, ...).

To watch the perf monitor latencies run from your build dir (e.g., for osd 0):
```
bin/ceph daemonperf osd.0
```

To watch measurements of latency and changes to throttle (e.g., for osd 0):
```
tail -f out/osd.0.log | grep throttle_mgr
```

Commands I used to generate requests:
```
bin/rados mkpool testp
bin/rados -b 4096 -o 4096 -t 8 --pool=testp bench 600 write
```

What I found is that for periods the latency (**clat** column on `daemonperf` command) was stable. But occasionally it would also spike for a while and even if the throttle was expanded, it would not bring down the latency.

I suspect that the throttle should be reduced while the submit latency (or throttle latency) is 0. I will work on that code next.

### CODE REVIEW

**Comment by ivancich** [comment]
Here's an example set of lines in ceph.conf:
        debug osd = 30
	bluestore latency managed throttle = true
	bluestore throttle latency min = 1250
	bluestore throttle latency max = 1750
	bluestore max bytes = 33554432 ; 32 * 1024 * 1024

**Comment by markhpc** [comment]
sounds good, hoping to start testing this today!

**Comment by markhpc** [comment]
Just an update on this, I tested both with the throttle on and throttle off, and saw lower performance, especially in the HDD data NVMe wal/db cases when doing sync writes at high IO depths.  wallclock profile showed much higher time spent in kerneldevice flush.  Not sure why, there are a couple of conditions there that can trigger a flush but I haven't dug into it further to figure out which condition is trigger the flush more often.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -3291,11 +3305,43 @@ void BlueStore::handle_conf_change(const struct md_config_t *conf,
...
       conf->bluestore_max_bytes + conf->bluestore_deferred_max_bytes);
+    // reset internal throttle of the throttle manager
+    throttle_mgr.reset();
   }
   if (changed.count("bluestore_deferred_max_bytes")) {
     throttle_deferred_bytes.reset_max(
       conf->bluestore_max_bytes + conf->bluestore_deferred_max_bytes);
   }
+
+#if 0
```
Comment (amitkumar50) [inline src/os/bluestore/BlueStore.cc]: @ivancich If not really required we can remove the code?

**Pull request closed without merging**